network-mux/src/Network/Mux.hs

```diff
@@ -157,26 +208,68 @@ muxStart tracer peerid app bearer = do
                               cnt
 
         return $ case app of
-          MuxInitiatorApplication initiator ->
-            [ ( initiator peerid mpdId initiatorChannel >> mpsJobExit cnt
+          MuxInitiatorApplication _ _ ->
+            [ ( mpsInitiatorBrooder itbl mpdId app initiatorChannel
               , show mpdId ++ " Initiator" )]
-          MuxResponderApplication responder ->
-            [ ( responder peerid mpdId responderChannel >> mpsJobExit cnt
-              , show mpdId ++ " Responder" )]
-          MuxInitiatorAndResponderApplication initiator responder ->
-            [ ( initiator peerid mpdId initiatorChannel >> mpsJobExit cnt
+          MuxResponderApplication _ _ ->
+            [ (mpsResponderBrooder pmss rtbl mpdId app responderChannel
+            , show mpdId ++ " Responder" )]
+          MuxInitiatorAndResponderApplication _ _ _ _ ->
+            [ ( mpsInitiatorBrooder itbl mpdId app initiatorChannel
               , show mpdId ++ " Initiator" )
-            , (responder peerid mpdId responderChannel >> mpsJobExit cnt
+            , (mpsResponderBrooder pmss rtbl mpdId app responderChannel
               , show mpdId ++ " Responder" )
             ]
 
-    -- cnt represent the number of SDUs that are queued but not yet sent.  Job
-    -- threads will be prevented from exiting until all SDUs have been
-    -- transmitted unless an exception/error is encounter. In that case all
-    -- jobs will be cancelled directly.
-    mpsJobExit :: StrictTVar m Int -> m ()
-    mpsJobExit cnt = do
+    -- Start and restart responder side miniprotocols on demand.
+    mpsResponderBrooder
+      :: HasResponder appType ~ True
+      => PerMuxSharedState ptcl m
+      -> MiniProtocolResponderControlTable ptcl m b
+      -> ptcl
+      -> MuxApplication appType peerid ptcl m a b
+      -> Channel m
+      -> m ()
+    mpsResponderBrooder pmss (MiniProtocolResponderControlTable rtbl) mpdId rspApp channel = forever $ do
+        -- Wait for initiator until we startup
+        traceWith tracer $ MuxTraceBrooderResponderBrood (AppProtocolId mpdId)
+        atomically $ do
+            buf <- readTVar (ingressQueue (dispatchTable pmss) (AppProtocolId mpdId) ModeResponder)
```

Comment:
You might have fixed this in one of the next commits: here we read from a queue and if there was something we ignore it, which will cause a problem when later we start the miniprotocol (likely a decoding failure, or a wrong message or a timeout error).

Reply:
It is a `readTVar` not `readTMvar`, no data is actually consumed.